src/v/security/license.cc

```diff
@@ -0,0 +1,302 @@
+    switch (type) {
+    case 0:
+        return license_type::free_trial;
+    case 1:
+        return license_type::enterprise;
+    default:
+        throw license_invalid_exception("Unknown license_type");
+    }
+}
+
+std::ostream& operator<<(std::ostream& os, const license& lic) {
+    fmt::print(os, "{}", lic);
+    return os;
+}
+
+struct license_components {
+    ss::sstring data;
+    ss::sstring signature;
+};
+
+static license_components parse_license(const ss::sstring& license) {
+    static constexpr auto signature_delimiter = ".";
+    const auto itr = license.find(signature_delimiter);
+    if (itr == ss::sstring::npos) {
+        throw license_malformed_exception("Outer envelope malformed");
+    }
+    /// signature encoded b64 first before encoding as utf-8 string, this is
+    /// done so that it can have a utf-8 interpretation so the license file
+    /// doesn't have to be in binary format
+    return license_components{
+      .data = license.substr(0, itr),
+      .signature = base64_to_string(
+        license.substr(itr + strlen(signature_delimiter)))};
+}
+
+static void parse_data_section(license& lc, const json::Document& doc) {
+    auto parse_int = [](auto& value) {
+        if (!value.IsInt()) {
+            throw license_malformed_exception("Bad cast: expected int");
+        }
+        return value.GetInt();
+    };
+    auto parse_str = [](auto& value) -> std::string {
+        if (!value.IsString()) {
+            throw license_malformed_exception("Bad cast: expected string");
+        }
+        return value.GetString();
+    };
+
+    auto parse_expiry = [&](auto& value) -> boost::gregorian::date {
+        auto expiry_str = parse_str(value);
+        boost::gregorian::date expiry_date;
+        try {
+            expiry_date = boost::gregorian::from_simple_string(expiry_str);
+        } catch (const boost::bad_lexical_cast& ex) {
+            throw license_malformed_exception(
+              fmt::format("Bad cast: Expiry dateformat: {}", ex.what()));
+        }
+        if (expiry_date.is_not_a_date()) {
+            throw license_malformed_exception(
+              "Expiration date not a real calendar date");
+        }
+        const auto today = boost::gregorian::day_clock::universal_day();
+        if (expiry_date < today) {
+            throw license_invalid_exception("Expiry date behind todays date");
+        }
+        return expiry_date;
+    };
+
+    static const std::array<ss::sstring, 4> v0_license_schema{
+      "version", "org", "type", "expiry"};
+
+    const size_t schema_keys_found = std::count_if(
+      doc.MemberBegin(), doc.MemberEnd(), [&](const auto& item) {
+          return std::find(
+                   v0_license_schema.begin(),
+                   v0_license_schema.end(),
+                   parse_str(item.name))
+                 != v0_license_schema.end();
+      });
+    if (schema_keys_found < v0_license_schema.size()) {
```

Comment:
consider using json schema for this. it is more robust than hand coding the verification. there is a short example of how to use it in c++ in admin_server.cc